File: slicing_layer/slicing.py
import numpy as np
from utils import read_bin, save_bin

def slicing_channel_fist(array):
    assert len(array.shape) == 3, "wrong input shape"

    #slicing here since delited it in the model
    patch_top_left = array[..., ::2, ::2]
    patch_top_right = array[..., ::2, 1::2]
    patch_bot_left = array[..., 1::2, ::2]
    patch_bot_right = array[..., 1::2, 1::2]
    array = np.concatenate(
        (
            patch_top_left,
            patch_bot_left,
            patch_top_right,
            patch_bot_right,
        ),
        axis=0,
    )
    return array

def slicing_channel_last(array):
    assert len(array.shape) == 3, "wrong input shape"

    patch_top_left = array[::2, ::2, :]
    patch_bot_left = array[1::2, ::2, :]
    patch_top_right = array[::2, 1::2, :]
    patch_bot_right = array[1::2, 1::2, :]
    array = np.concatenate(
        (
            patch_top_left,
            patch_bot_left,
            patch_top_right,
            patch_bot_right,
        ),
        axis=-1,
    )
    return array

def slicing_chw_c_style(array, h, w, chnl):

    cout = 0 
    tmp = array.copy()
    # top left
    for c in range(0, chnl):
        for j in range(0, h // 2):
            k = j * (2 * w)  
            for i in range(0, w // 2): 
                tmp[cout] = array[k + i * 2 + c *(w  * h)]
                cout += 1

    # bottom left
    for c in range(0, chnl):
        # print("c = ", c )
        for j in range(0, h // 2):
            k = j * (2 * w) + w  
            for i in range(0, w // 2): 
                tmp[cout] = array[k + i * 2 + c *(w  * h)]
                cout += 1

    # top right 
    for c in range(0, chnl):
        for j in range(0, h // 2):
            k = j * (2 * w) + 1  
            for i in range(0, w // 2): 
                tmp[cout] = array[k + i * 2 + c * (w * h)]
                cout += 1

    # bottom right
    for c in range(0, chnl):
        for j in range(0, h // 2):
            k = j * (2 * w) + w + 1 
            for i in range(0, w // 2): 
                tmp[cout] = array[k + i * 2 + c * (w * h)]
                cout += 1
    return tmp


def slicing_hwc_c_style(array, h, w, channels):

    tmp_array = array.copy()
    cur = 0
    tmp1 = 0 
    tmp2 = w * channels 
    comp = (w * h * channels) // (h // 2) 

    larger, smaller = (h, w) if h > w else (w, h)

    for j in range(0, larger):
        for i in range(0, smaller):
            for c in range(0, channels):
                if i % 2 == 0:                
                    tmp_array[cur] = array[tmp1]
                    tmp1 += 1
                else:
                    tmp_array[cur] = array[tmp2]
                    tmp2 += 1 

                cur += 1
                if  cur % comp == 0: 
                    tmp1 += w * channels
                    tmp2 += w * channels
                
    return tmp_array


def slicing_hwc_c_style_imp(array, h, w, channels):

    o_h, o_w = h // 2, w // 2
    o_c = channels * 4
    output = np.zeros((o_h * o_w * o_c ), dtype=np.uint8)

    for j in range(0, o_h):
        for i in range(0, o_w):
            for c in range(0, channels): 
                output[j * o_w * o_c + i * o_c + 0 * channels + c] = array[(j * 2 * w * channels) + (i * 2 * channels) + c]
                output[j * o_w * o_c + i * o_c + 1 * channels + c] = array[(j * 2 * w * channels) + (i * 2 * channels) + (channels * w) +  c]
                output[j * o_w * o_c + i * o_c + 2 * channels + c] = array[(j * 2 * w * channels) + (i * 2 * channels) + channels  + c]
                output[j * o_w * o_c + i * o_c + 3 * channels + c] = array[(j * 2 * w * channels) + (i * 2 * channels) + (channels * w + channels)  + c]
                # output[j * o_w * o_c + i * o_c + c]     = array[(j * 2 * w * channels) + (i * 2 * channels) + c]
                # output[j * o_w * o_c + i * o_c + 3 + c] = array[(j * 2 * w * channels) + (i * 2 * channels) + (channels * w) +  c]
                # output[j * o_w * o_c + i * o_c + 6 + c] = array[(j * 2 * w * channels) + (i * 2 * channels) + channels  + c]
                # output[j * o_w * o_c + i * o_c + 9 + c] = array[(j * 2 * w * channels) + (i * 2 * channels) + (channels * w + channels)  + c]
    return output 

def check_idex(input_array, input_seq, h, w, channels): 

    o_h, o_w = h // 2, w // 2
    o_c = channels * 4
    output = np.zeros((o_h, o_w, o_c), dtype=np.uint8)
    output_seq = np.zeros((input_seq.shape), dtype=np.uint8)

    for j in range(0, o_h):
        for i in range(0, o_w):
            for c in range(0, channels): 

                output[j, i, 0 * channels + c] = input_array[j * 2, i * 2, c]
                output[j, i, 1 * channels + c] = input_array[j * 2 + 1, i * 2, c]
                output[j, i, 2 * channels + c] = input_array[j * 2, i * 2 + 1, c]
                output[j, i, 3 * channels + c] = input_array[1 + j * 2, 1 + i * 2, c]
            
                output_seq[j * o_w * o_c + i * o_c + 0 * channels + c] = input_seq[(j * 2 * w * channels) + (i * 2 * channels) + c]
                output_seq[j * o_w * o_c + i * o_c + 1 * channels + c] = input_seq[(j * 2 * w * channels) + (i * 2 * channels) + (channels * w) +  c]
                output_seq[j * o_w * o_c + i * o_c + 2 * channels + c] = input_seq[(j * 2 * w * channels) + (i * 2 * channels) + channels  + c]
                output_seq[j * o_w * o_c + i * o_c + 3 * channels + c] = input_seq[(j * 2 * w * channels) + (i * 2 * channels) + (channels * w + channels)  + c]

    input_seq = input_seq.reshape((h, w, channels))
    output_seq = output_seq.reshape((o_h, o_w, o_c))

    cor_output = slicing_channel_last(input_array)
    print(np.sum(output - output_seq))

def slicing_hw1_style(array, h, w):

    tmp_array = array.copy()
    cur = 0
    tmp1 = 0 
    tmp2 = w 
    comp = (w * h) // (h // 2) 
    larger, smaller = (h, w) if h > w else (w, h)

    for j in range(0, larger):
        for i in range(0, smaller):
            if i % 2 == 0:                
                tmp_array[cur] = array[tmp1]
                tmp1 += 1
            else:
                tmp_array[cur] = array[tmp2]
                tmp2 += 1 

            cur += 1
            if  cur % comp == 0: 
                tmp1 += w
                tmp2 += w

    return tmp_array

if __name__ == "__main__":

    np.random.seed(1)

    h = 240  
    w = 320 
    c = 3
    array = np.arange(h * w * c, dtype=np.uint64).reshape(h, w, c)
    # array = np.random.randint(low=0, high=255, size=(h * w * c), dtype=np.uint8).reshape(h, w, c )
    
    # test the dataset on the slicing function
    h, w, c = array.shape
    print(array)
    print(array.shape)

    save_bin(array, "./array.bin")
    array_python = read_bin("./array.bin")

    print(array_python)
    print(array_python.shape)

    array_scliced_python = slicing_channel_last(array)

    print(array_scliced_python)
    print(array_scliced_python.shape)

    save_bin(array_scliced_python, "array_scliced_python.bin")
    array_sliced_bin = read_bin("array_scliced_python.bin")


    print(array_sliced_bin)
    print(array_sliced_bin.shape)

    # call the fucntion 
    sliced_c_stype = slicing_hwc_c_style_imp(array_python, h, w, channels=c) 

    mistach = np.sum(sliced_c_stype - array_sliced_bin)
    print(f"mistach = {mistach} w = {w} h = {h}")
    assert mistach == 0, f"mismatch: {mistach} h={h} w={w} c={c}"
            

File: slicing_layer/slicing.c
#include <stdio.h>
#include <stdlib.h> 

void slicing_hwc_channel(signed char * array, int h, int w, int channels){

    signed char * tmp = malloc(h * w * channels * sizeof(signed char));

    unsigned int cur = 0; 
    unsigned int tmp1 = 0;
    unsigned int tmp2 = w * channels;
    unsigned int comp = (w * h * channels) / (h / 2);

    int larger, smaller;
    if(h > w){
        larger = h;
        smaller = w;
    }
    else{
        larger = w;
        smaller = h;
    }

    for(int j = 0; j < larger; j++){
        for(int i = 0; i < smaller; i++){
            for(int c = 0; c < channels; c++){
                if (i % 2 == 0){
                    tmp[cur] = array[tmp1];
                    tmp1++;
                }else{
                    tmp[cur] = array[tmp2];
                    tmp2++;
                }
                cur++;
                if (cur % comp == 0){
                    tmp1 += w * channels;
                    tmp2 += w * channels;
                }
            }
        }
    }
    
    for(int i = 0; i < h * w * channels; i++){
        array[i] = tmp[i];
    }
    free(tmp);
}

void slicing_chw_channel(signed char * array, int h, int w, int chnls){

    signed char * tmp = malloc(h * w * chnls * sizeof(signed char));

    for(int i = 0; i < h * w * chnls; i++){
        tmp[i] = array[i];
    }

    unsigned int count = 0;
    unsigned int k = 0;

    // top left 
    for(int c = 0; c < chnls; c++){
        for(int j = 0; j < h / 2; j++){
            k = j * ( 2 * w);
            for(int i=0; i < w / 2; i ++){
                tmp[count] = array[k + i * 2 + c * (w * h)];
                count++;
            }
        }
    }


    // bottom left 
    for(int c = 0; c < chnls; c++){
        for(int j = 0; j < h / 2; j++){
            k = j * ( 2 * w) + w;
            for(int i=0; i < w / 2; i ++){
                tmp[count] = array[k + i * 2 + c * (w * h)];
                count++;
            }
        }
    }

    // top right 
    for(int c = 0; c < chnls; c++){
        for(int j = 0; j < h / 2; j++){
            k = j * ( 2 * w) + 1;
            for(int i=0; i < w / 2; i ++){
                tmp[count] = array[k + i * 2 + c * (w * h)];
                count++;
            }
        }
    }

    // bottom right 
    for(int c = 0; c < chnls; c++){
        for(int j = 0; j < h / 2; j++){
            k = j * (2 * w) + w +  1;
            for(int i=0; i < w / 2; i ++){
                tmp[count] = array[k + i * 2 + c * (w * h)];
                count++;
            }
        }
    }

    for(int i = 0; i < h * w * chnls; i++){
        array[i] = tmp[i];
    }
    free(tmp);
}

void slicing(signed char * array, unsigned char * tmp, int h, int w){

    
    for(int i = 0; i < h * w; i++){
        tmp[i] = array[i];
    }

    unsigned int count = 0;

    // top left 
    for(int j = 0; j < h / 2; j++){
        unsigned int k = j * ( 2 * w);
        // printf("k ==: %d \n", k);
        for(int i=0; i < w / 2; i ++){
            tmp[count] = array[k + i * 2];
            count++;
        }
    }

    // bottom left 
    for(int j = 0; j < h / 2; j++){
        unsigned int k = j * ( 2 * w) + w;
        // printf("k ==: %d \n", k);
        for(int i=0; i < w / 2; i ++){
            tmp[count] = array[k + i * 2];
            count++;
        }
    }

    // top right 
    for(int j = 0; j < h / 2; j++){
        unsigned int k = j * ( 2 * w) + 1;
        // printf("k ==: %d \n", k);
        for(int i=0; i < w / 2; i ++){
            tmp[count] = array[k + i * 2];
            count++;
        }
    }

    // bottom right 
    for(int j = 0; j < h / 2; j++){
        unsigned int k = j * (2 * w) + w +  1;
        // printf("k ==: %d \n", k);
        for(int i=0; i < w / 2; i ++){
            tmp[count] = array[k + i * 2];
            count++;
        }
    }
    
    for(int i = 0; i < h * w; i++){
        array[i] = tmp[i];
    }
}

// int main(){

//     int h = 2;
//     int w = 4;
//     int c = 3;
//     unsigned int size_buff = h * w * c; 
//     unsigned int size = h * w * c;
//     unsigned char buffer[size_buff];
//     unsigned char target[size_buff];

//     printf("read it ");
//     FILE *ptr;
//     ptr = fopen("./array_2_4_3.bin", "rb");
//     fread(buffer, sizeof(buffer), 1, ptr);

//     FILE *ptr1;
//     ptr1 = fopen("./array_scliced_2_4_3.bin", "rb");
//     fread(target, sizeof(target), 1, ptr1);
    
//     printf("================== data \n");
//     for(int i = 0; i < size; i++)
//         printf("%d ", buffer[i]); // prints a series of bytes
//     printf("\n");

//     printf("================== target \n");
//     for(int i = 0; i < size; i++)
//         printf("%d ", target[i]); // prints a series of bytes
//     printf("\n");
//     printf("================== \n");

//     // slicing_3_channel(buffer, 30, 242, 3);
//     slicing_hwc_channel(buffer, h, w, c);


//     printf("================== sliced in C \n");
//     for(int i = 0; i < size; i++)
//         printf("%d ", buffer[i]); // prints a series of bytes
//     printf("\n");

//     unsigned mismatch = 0;

//     printf("================== mismatch\n");
//     for(int i = 0; i < size_buff; i++)
//         mismatch += target[i] - buffer[i];

//     printf("mismatch = %u ", mismatch); // prints a series of bytes
//     printf("\n");

// }